src/fparser/two/tests/fortran2003/test_char_literal_constant_r427.py

```diff
@@ -0,0 +1,126 @@
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+'''Test Fortran 2003 rule R427 : This file tests the support for a
+character literal constant.
+
+'''
+
+import pytest
+from fparser.two.Fortran2003 import Char_Literal_Constant
+from fparser.two.utils import NoMatchError, InternalError
+
+
+def test_match_valid():
+    ''' Test that valid input is parsed correctly '''
+
+    # simple, single quotes
+    obj = Char_Literal_Constant("'DO'")
+    assert isinstance(obj, Char_Literal_Constant), repr(obj)
+    assert str(obj) == "'DO'"
+    assert repr(obj) == 'Char_Literal_Constant("\'DO\'", None)'
+
+    # simple, double quotes
+    obj = Char_Literal_Constant('"DO"')
+    assert isinstance(obj, Char_Literal_Constant), repr(obj)
+    assert str(obj) == '"DO"'
+
+    # single quotes inside single quotes (two means one)
+    obj = Char_Literal_Constant("'DON''T'")
+    assert isinstance(obj, Char_Literal_Constant), repr(obj)
+    assert str(obj) == "'DON''T'"
+
+    # double quotes inside double quotes (two means one)
+    obj = Char_Literal_Constant('"""busy"""')
+    assert isinstance(obj, Char_Literal_Constant), repr(obj)
+    assert str(obj) == '"""busy"""'
+
+    # single quotes, spaces
+    obj = Char_Literal_Constant("  '  D  O  '  ")
+    assert isinstance(obj, Char_Literal_Constant), repr(obj)
+    assert str(obj) == "'  D  O  '"
+    assert repr(obj) == 'Char_Literal_Constant("\'  D  O  \'", None)'
+
+    # Single quotes, empty string
+    obj = Char_Literal_Constant("''")
+    assert isinstance(obj, Char_Literal_Constant), repr(obj)
+    assert str(obj) == "''"
+
+    # Double quotes, empty string
+    obj = Char_Literal_Constant('""')
+    assert isinstance(obj, Char_Literal_Constant), repr(obj)
+    assert str(obj) == '""'
+
+    # include a kind parameter (which says what character set to
+    # expect)
+    obj = Char_Literal_Constant('KP_"DO"')
+    assert isinstance(obj, Char_Literal_Constant), repr(obj)
+    assert str(obj) == 'KP_"DO"'
+    assert repr(obj) == 'Char_Literal_Constant(\'"DO"\', \'KP\')'
+
+    # include a kind parameter with spaces
+    obj = Char_Literal_Constant('  KP  _  "  D  O  "  ')
+    assert isinstance(obj, Char_Literal_Constant), repr(obj)
+    assert str(obj) == 'KP_"  D  O  "'
+    assert repr(obj) == 'Char_Literal_Constant(\'"  D  O  "\', \'KP\')'
+
+    # additional characters
+    obj = Char_Literal_Constant("'()!$%^&*_+=-01~@#;:/?.>,<|'")
+    assert isinstance(obj, Char_Literal_Constant), repr(obj)
+    assert str(obj) == "'()!$%^&*_+=-01~@#;:/?.>,<|'"
+
+
+def test_match_invalid():
+    ''' Test that invalid input raises an exception '''
+
+    # test various invalid options
+    for example in ["", "  ", "A", "'A", "A'", "\"A", "A\"", "A'A'", "A 'A'",
```

Comment:
If i'm reading this correctly, this says that an empty string in double quotes is an error but earlier we have a test that says an empty string with single quotes is valid?

Reply:
Also, why is "A" an error (or are those quotes not double quotes)?

Reply:
That's annoying, I did write something so I must have deleted it somehow. Anyway ...

The input string is expected to include quotes, so "A" and 'A' are invalid as the string itself does not have quotes. "'A'" or '"A"' would be valid, as would "''" or '""'.